src/Functions/geometryConverters.h

```diff
@@ -0,0 +1,547 @@
+        : col(col_)
+        , offsets(dynamic_cast<const ColumnArray &>(*col_).getOffsets())
+        , point_parser(dynamic_cast<const ColumnArray &>(*col_).getDataPtr())
+    {
+    }
+
+    Geometry<Point> createContainer() const
+    {
+        return Ring<Point>();
+    }
+
+    void get(Geometry<Point> & container, size_t i) const
+    {
+        get(boost::get<Ring<Point>>(container), i);
+    }
+
+    void get(Ring<Point> & container, size_t i) const
+    {
+        size_t left = i == 0 ? 0 : offsets[i - 1];
+        size_t right = offsets[i];
+
+        if (left == right)
+            throw Exception("Empty polygons are not allowed in line " + toString(i), ErrorCodes::BAD_ARGUMENTS);
+
+        // reserve extra point for case when polygon is open
+        container.reserve(right - left + 1);
+        container.resize(right - left);
+
+        for (size_t j = left; j < right; j++)
+            point_parser.get(container[j - left], j);
+
+        // make ring closed
+        if (!boost::geometry::equals(container[0], container.back()))
+        {
+            container.push_back(container[0]);
+        }
+    }
+
+private:
+    /// To prevent use-after-free and increase column lifetime.
+    ColumnPtr col;
+    const IColumn::Offsets & offsets;
+    const PointFromColumnParser<Point> point_parser;
+};
+
+template<class Point>
+class PolygonFromColumnParser
+{
+public:
+    explicit PolygonFromColumnParser(ColumnPtr col_)
+        : col(col_)
+        , offsets(static_cast<const ColumnArray &>(*col_).getOffsets())
+        , ring_parser(static_cast<const ColumnArray &>(*col_).getDataPtr())
+    {}
+
+    Geometry<Point> createContainer() const
+    {
+        return Polygon<Point>();
+    }
+
+    void get(Geometry<Point> & container, size_t i) const
+    {
+        get(boost::get<Polygon<Point>>(container), i);
+    }
+
+    void get(Polygon<Point> & container, size_t i) const
+    {
+        size_t l = offsets[i - 1];
+        size_t r = offsets[i];
+
+        ring_parser.get(container.outer(), l);
+
+        container.inners().resize(r - l - 1);
+        for (size_t j = l + 1; j < r; j++)
+        {
+            ring_parser.get(container.inners()[j - l - 1], j);
+        }
+    }
+
+private:
+    /// To prevent use-after-free and increase column lifetime.
+    ColumnPtr col;
+    const IColumn::Offsets & offsets;
+    const RingFromColumnParser<Point> ring_parser;
+};
+
+template<class Point>
+class MultiPolygonFromColumnParser
+{
+public:
+    explicit MultiPolygonFromColumnParser(ColumnPtr col_)
+        : col(col_)
+        , offsets(static_cast<const ColumnArray &>(*col_).getOffsets())
+        , polygon_parser(static_cast<const ColumnArray &>(*col_).getDataPtr())
+    {}
+
+    Geometry<Point> createContainer() const
+    {
+        return MultiPolygon<Point>();
+    }
+
+    void get(Geometry<Point> & container, size_t i) const
+    {
+        get(boost::get<MultiPolygon<Point>>(container), i);
```

Comment:
It will throw exception if container does not contain MultiPolygon<Point>.

Reply:
Deleted this code.